File: src/Enterprise/C#/BuilderExample.cs
using System;
using System.Collections.Generic;

public interface IReportBuilder
{
    void Reset();
    void AddTitle(string title);
    void AddSection(string heading, string body);
    string Build();
}

public sealed class TextReportBuilder : IReportBuilder
{
    private readonly List<string> _parts = new();

    public void Reset() => _parts.Clear();

    public void AddTitle(string title) => _parts.Add($"# {title}");

    public void AddSection(string heading, string body)
    {
        _parts.Add($"## {heading}");
        _parts.Add(body);
    }

    public string Build() => string.Join(Environment.NewLine, _parts);
}

public sealed class HtmlReportBuilder : IReportBuilder
{
    private readonly List<string> _parts = new();

    public void Reset() => _parts.Clear();

    public void AddTitle(string title) => _parts.Add($"<h1>{title}</h1>");

    public void AddSection(string heading, string body)
    {
        _parts.Add($"<h2>{heading}</h2>");
        _parts.Add($"<p>{body}</p>");
    }

    public string Build() => string.Join(string.Empty, _parts);
}

public static class ReportDirector
{
    public static string BuildAvailabilityReport(IReportBuilder builder)
    {
        builder.Reset();
        builder.AddTitle("Service status");
        builder.AddSection("Availability", "99.95%");
        return builder.Build();
    }
}

public static class BuilderExample
{
    public static void Main()
    {
        Console.WriteLine(ReportDirector.BuildAvailabilityReport(new TextReportBuilder()));
        Console.WriteLine("---");
        Console.WriteLine(ReportDirector.BuildAvailabilityReport(new HtmlReportBuilder()));
    }
}

File: src/Enterprise/VisualBasic/BuilderExample.vb
Imports System
Imports System.Collections.Generic

Public Interface IReportBuilder
    Sub Reset()
    Sub AddTitle(title As String)
    Sub AddSection(heading As String, body As String)
    Function Build() As String
End Interface

Public Class TextReportBuilder
    Implements IReportBuilder
    Private ReadOnly parts As New List(Of String)()

    Public Sub Reset() Implements IReportBuilder.Reset
        parts.Clear()
    End Sub

    Public Sub AddTitle(title As String) Implements IReportBuilder.AddTitle
        parts.Add("# " & title)
    End Sub

    Public Sub AddSection(heading As String, body As String) Implements IReportBuilder.AddSection
        parts.Add("## " & heading)
        parts.Add(body)
    End Sub

    Public Function Build() As String Implements IReportBuilder.Build
        Return String.Join(Environment.NewLine, parts)
    End Function
End Class

Public Class HtmlReportBuilder
    Implements IReportBuilder
    Private ReadOnly parts As New List(Of String)()

    Public Sub Reset() Implements IReportBuilder.Reset
        parts.Clear()
    End Sub

    Public Sub AddTitle(title As String) Implements IReportBuilder.AddTitle
        parts.Add("<h1>" & title & "</h1>")
    End Sub

    Public Sub AddSection(heading As String, body As String) Implements IReportBuilder.AddSection
        parts.Add("<h2>" & heading & "</h2>")
        parts.Add("<p>" & body & "</p>")
    End Sub

    Public Function Build() As String Implements IReportBuilder.Build
        Return String.Concat(parts)
    End Function
End Class

Public Module BuilderExample
    Private Function BuildAvailabilityReport(builder As IReportBuilder) As String
        builder.Reset()
        builder.AddTitle("Service status")
        builder.AddSection("Availability", "99.95%")
        Return builder.Build()
    End Function

    Public Sub Main()
        Console.WriteLine(BuildAvailabilityReport(New TextReportBuilder()))
        Console.WriteLine("---")
        Console.WriteLine(BuildAvailabilityReport(New HtmlReportBuilder()))
    End Sub
End Module
